File: 120.triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        vector<int> sums = triangle[0];
        for (int i = 1; i < triangle.size(); i++) {
            vector<int> row = triangle[i];
            int len = row.size();
            vector<int> new_sums(len);
            new_sums[0] = sums[0] + row[0];
            for (int j = 1; j < len - 1; j++) {
                new_sums[j] = min(sums[j - 1], sums[j]) + row[j];
            }
            new_sums[len - 1] = sums[len - 2] + row[len - 1];
            sums = new_sums;
        }
        return *min_element(sums.begin(), sums.end());
    }
};

File: 120.triangle.py
class Solution:
    def minimumTotal(self, triangle: 'List[List[int]]') -> 'int':
        sums = triangle[0];
        for i in range(1, len(triangle)):
            row = triangle[i]
            new_sums = [0] * len(row)
            new_sums[0] = sums[0] + row[0]
            for j in range(1, len(row) - 1):
                new_sums[j] = min(sums[j - 1], sums[j]) + row[j]
            new_sums[-1] = sums[-1] + row[-1]
            sums = new_sums
        return min(sums)
